### PR TITLE
Operator diagnostics mechanism

### PR DESCRIPTION
**Pull request by szalpal**
Operator diagnostics mechanism

Signed-off-by: Michal Szolucha <mszolucha@nvidia.com>

### CODE REVIEW

**Comment by szalpal** [comment]
!build

File: dali/pipeline/operator/operator.h [hunk 1]
```diff
@@ -166,6 +168,26 @@ class DLL_PUBLIC OperatorBase {
...
+  T GetDiagnostic(const std::string &name) const {
+    return *any_cast<T *>(counters_.at(name));
+  }
+
+  template<typename T>
+  void RegisterDiagnostic(std::string name, T *val) {
+    using namespace std;  // NOLINT
+    static_assert(is_arithmetic_or_half<remove_reference_t<T>>::value || is_same<T, string>::value,
+                  "The eligible counter types are arithmetic types or string");
+    // Note the WAR for `any` bug - when T* is passed to
```
Comment (mzient) [inline dali/pipeline/operator/operator.h:182]: Fix `any` instead - change the universal reference constructor to the following:
```c++
  template <typename T>
  any(T &&value) {  // NOLINT
    using U = std::remove_reference_t<T>;
    static_assert(std::is_copy_constructible<U>::value,
      "only copy-constructible types can be stored in 'any'");
    static_assert(std::is_destructible<U>::value,
      "objects stored in 'any' must be destructible");
    assign<U>(value);
  }
```

File: dali/pipeline/operator/operator.h [hunk 2]
```diff
@@ -166,6 +168,26 @@ class DLL_PUBLIC OperatorBase {
...
+    // forwarding reference ctor, T*& gets instantiated,
+    // which is a problematic type.
+    if (!counters_.emplace(move(name), (val + 0)).second) {
+      DALI_FAIL("Counter with given name already exists");
```
Comment (mzient) [inline dali/pipeline/operator/operator.h:186]:
```suggestion
      DALI_FAIL("Diagnostic entry with given name already exists");
```

File: dali/pipeline/operator/operator.h [hunk 3]
```diff
@@ -217,10 +239,13 @@ class DLL_PUBLIC OperatorBase {
     assert(output.size() == static_cast<size_t>(batch_size_));
   }
 
+
   const OpSpec spec_;
   int num_threads_;
   int batch_size_;
   int default_cuda_stream_priority_;
+
+  std::unordered_map<std::string, any> counters_;
```
Comment (mzient) [inline dali/pipeline/operator/operator.h:248]:
```suggestion
  std::unordered_map<std::string, any> diagnostics_;
```
?

Comment (szalpal) [reply dali/pipeline/operator/operator.h:248]: Yeah, I missed it. Thanks for the catch

Comment (szalpal) [reply dali/pipeline/operator/operator.h]: Done

File: dali/pipeline/operator/operator.h (hunk 2 above)
Comment (szalpal) [reply dali/pipeline/operator/operator.h]: Done

File: dali/pipeline/operator/operator.h (hunk 1 above)
Comment (szalpal) [reply dali/pipeline/operator/operator.h]: Done

File: dali/pipeline/operator/operator.h [hunk 4]
```diff
@@ -166,6 +168,23 @@ class DLL_PUBLIC OperatorBase {
...
+
+  template<typename T>
+  T GetDiagnostic(const std::string &name) const {
+    return *any_cast<T *>(diagnostics_.at(name));
+  }
+
+  template<typename T>
+  void RegisterDiagnostic(std::string name, T *val) {
+    using namespace std;  // NOLINT
+    static_assert(is_arithmetic_or_half<remove_reference_t<T>>::value || is_same<T, string>::value,
```
Comment (mzient) [inline dali/pipeline/operator/operator.h:180]: I'd say that the limitation should be to POD types. `string` is dangerous - it's a complex type with internal dynamic storage and requirements for consistency - unguarded parallel access may result in fatal race conditions.

Comment (szalpal) [reply dali/pipeline/operator/operator.h]: removed string, added enum

File: dali/pipeline/operator/operator.h [hunk 5]
```diff
@@ -166,6 +168,23 @@ class DLL_PUBLIC OperatorBase {
...
+  template<typename T>
+  T GetDiagnostic(const std::string &name) const {
+    return *any_cast<T *>(diagnostics_.at(name));
+  }
+
+  template<typename T>
+  void RegisterDiagnostic(std::string name, T *val) {
+    using namespace std;  // NOLINT
+    static_assert(is_arithmetic_or_half<remove_reference_t<T>>::value || is_enum<T>::value,
+                  "The eligible counter types are arithmetic types or string");
```
Comment (JanuszL) [inline dali/pipeline/operator/operator.h:181]:
```suggestion
                  "The eligible diagnostic entry types are arithmetic types or enum");
```

**Comment by szalpal** [comment]
!build

**Pull request merged**